src/common.rs

```diff
@@ -319,3 +413,20 @@ impl From<Utf8Error> for GetError {
         GetError
     }
 }
+
+/// Set keys to null where the union member is null.
+///
+/// This is a workaround to https://github.com/apache/arrow-rs/issues/6017#issuecomment-2352756753
+/// - i.e. that dictionary null is most reliably done if the keys are null.
+fn mask_dictionary_keys<K: ArrowPrimitiveType>(keys: &PrimitiveArray<K>, type_ids: &[i8]) -> PrimitiveArray<K> {
+    let mut null_builder = BooleanBufferBuilder::new(keys.len());
```

Comment:
It [seems](https://github.com/apache/datafusion/pull/12460#discussion_r1759705907) `BooleanBufferBuilder` is slower than `Vec<boo>` or `Vec<Option<bool>>`, especially as you can build the vec with `true` then only set the `false` values.

Reply:
Used a vec 👍